spec/activerecord-bitemporal/date_type_valid_time_spec.rb

```diff
@@ -0,0 +1,1112 @@
+            it { expect(cop1.valid_from).to eq cop2.valid_from }
+          end
+
+          context "department with valid_from" do
+            let(:department) { Department.new(name: "Department", valid_from: "2018/12/25") }
+            subject { department.valid_from }
+            before { department.save }
+
+            it { is_expected.to eq "2018/12/25".to_date }
+            it { is_expected.to eq cop1.valid_from }
+            it { is_expected.to eq cop2.valid_from }
+            it { expect(cop1.valid_from).to eq cop2.valid_from }
+          end
+        end
+
+        context "#create to #save" do
+          let(:department) { Timecop.travel(1.day.ago) { Department.create(name: "Department") } }
+
+          context "saving" do
+            subject { department.save }
+            it { expect { subject }.to change { cop1.valid_from } }
+            it { expect { subject }.to change { cop2.valid_from } }
+            it { expect { subject }.not_to change { cop3.valid_from } }
+          end
+
+          context "saved" do
+            before { department.save }
+            subject { department.valid_from }
+            it { is_expected.not_to eq cop1.valid_from }
+            it { is_expected.not_to eq cop2.valid_from }
+            it { expect(cop1.valid_from).to eq cop2.valid_from } # same date
+          end
+        end
+      end
+
+      describe "inverse_of" do
+        let(:department) do
+          klass = Class.new(Department) {
+            has_many :cops, foreign_key: :department_id, inverse_of: :department
+
+            def self.name
+              "DepartmentWithInverseOf"
+            end
+          }
+          Timecop.travel(1.day.ago) { klass.create!(name: "Department") }
+        end
+        before do
+          department.update(name: "Department2")
+          department.cops.create!(name: "Jane")
+        end
+
+        it { expect(department.reload.cops.first.department).to be department }
+      end
+    end
+
+    describe "Date type valid time has many Datetime(timestamp) type vaild time" do
+      describe "sync valid_from in create" do
+        let!(:manager1) { department.managers.new(name: "Jane") }
+        let!(:manager2) { department.managers.new(name: "Homu") }
+        let!(:manager3) { department.managers.new(name: "Mami", valid_from: "2016/01/15") }
+
+        context "#new to #save" do
+          let(:department) { Department.new(name: "Department") }
+
+          context "saving" do
+            subject { department.save }
+            it { expect { subject }.to change { manager1.valid_from } }
+            it { expect { subject }.to change { manager2.valid_from } }
+            it { expect { subject }.not_to change { manager3.valid_from } }
+          end
+
+          context "saved" do
+            subject { department.valid_from }
+            before { department.save }
+            it { is_expected.to eq manager1.valid_from.in_time_zone.to_date }
```

Comment:
This may seem a bit weird behavior. Presumably the value the you expects is the current time, not 00:00:00.

Reply:
Indeed, the intuitively expected value is the current time. However, the time information is lost because of the `valid_at!(self.valid_from)` in the following code.
https://github.com/kufu/activerecord-bitemporal/blob/374c72ffa6c7a4daeba59113bf057089b093c496/lib/activerecord-bitemporal/bitemporal.rb#L296

This is likely to be a limitation when using associations of different column types.